### DISCUSSION TITLE
.first() behavior

### DISCUSSION THREAD

**Discussion by sondrelg** [collerek/ormar#427] [Q&A]
Hi!

I'm trying to get the latest item from a table, wrt. a datetime-field, and I was a little surprised to learn how the `first` method works.

### What I expected

I've used Django a bit, so I was expecting .first() to get the first item in my queryset, with respect to existing constaints. I had already ordered the queryset the way
I wanted, so I was expecting this to work:

```python
latest = await MyModel.objects.order_by("-registered").first()
```

When that didn't work, I thought I might try to limit the size of the queryset myself, then use `.first()` as a shortcut over `.all()` since this will raise `NoMatch` etc. when the object doesn't exist:

```python
latest = await MyModel.objects.order_by("-registered").limit(1).first()
```

### Question

Is it correct that `.first()` disregards the queryset it's being called on and just returns the first item, sorted by pk? Why is this? Am I missing alternative syntax for a Django-style `.first()` method?

This is what I ended up with:

```python
latest = await MyModel.objects.order_by("-registered").limit(1).all()

if not latest:
    # execute the logic I'd otherwise have defined in the NoMatch case

return ...
```

Thanks!

**Comment by collerek** [comment]
That might be a bug :)
The first() and get() behaviour changes depending on if you provided the filters or not, but I think I didn't include the possibility to change the behaviour based on ordering.

I always prepend primary key ordering in that case:
```python
expr = self.build_select_expression(
    limit=1,
    order_bys=[
      OrderAction(
          order_str=f"{self.model.Meta.pkname}",
          model_cls=self.model_cls, 
      )
    ]
    + self.order_bys,
    )
```

I guess that should happen only if you do not provide any ordering for the main model. 
There have to be ordering on the main model as you need to fetch the 1 row of the master model but all child models of the master model so you cannot simply use SQL limit/offset clauses as one master model will have multiple rows after joins and I don't want to issue a separate query to check the number of the rows. Can you open an issue for this, will check this out.

**Reply by sondrelg** [reply]
Yep, will do, thanks!
